### PR TITLE
Enable gp_recursive_cte during statement preparation

### PR DESCRIPTION
**Pull request by KonstantAnxiety**

### CODE REVIEW

File: lib/dl_connector_greenplum/dl_connector_greenplum/core/adapters.py [hunk 1]
```diff
@@ -72,3 +77,20 @@ class AsyncGreenplumAdapter(AsyncPostgresAdapter):
...
+            db_version = await connection.fetchval("SELECT version()")
+            if "greenplum" not in db_version.lower():
+                # to keep compatibility with postgres
+                yield
+            else:
+                gp_recursive_cte_initial = await connection.fetchval("SHOW gp_recursive_cte")
+                if gp_recursive_cte_initial == "off":
+                    await connection.execute("SET gp_recursive_cte = on")
+
+                yield
```
Comment (ovsds) [inline lib/dl_connector_greenplum/dl_connector_greenplum/core/adapters.py:93]: try:
  yield
finally:
  ...

File: lib/dl_connector_greenplum/dl_connector_greenplum/core/adapters.py [hunk 2]
```diff
@@ -72,3 +77,20 @@ class AsyncGreenplumAdapter(AsyncPostgresAdapter):
 
     async def get_table_info(self, table_def: TableDefinition, fetch_idx_info: bool) -> RawSchemaInfo:
         raise NotImplementedError()
+
+    @asynccontextmanager
+    async def _query_preparation_context(self, connection: asyncpg.Connection) -> AsyncIterator[None]:
+        async with super()._query_preparation_context(connection):
+            db_version = await connection.fetchval("SELECT version()")
+            if "greenplum" not in db_version.lower():
```
Comment (ovsds) [inline lib/dl_connector_greenplum/dl_connector_greenplum/core/adapters.py:85]: do we really need it in GP Adapter?

File: lib/dl_connector_greenplum/dl_connector_greenplum/core/adapters.py [hunk 3]
```diff
@@ -72,3 +77,20 @@ class AsyncGreenplumAdapter(AsyncPostgresAdapter):
...
+                yield
+            else:
+                gp_recursive_cte_initial = await connection.fetchval("SHOW gp_recursive_cte")
+                if gp_recursive_cte_initial == "off":
+                    await connection.execute("SET gp_recursive_cte = on")
+
+                yield
+
+                if gp_recursive_cte_initial == "off":
+                    await connection.execute("SET gp_recursive_cte = off")
```
Comment (ovsds) [inline lib/dl_connector_greenplum/dl_connector_greenplum/core/adapters.py:96]: nitpick: could be restructered for flatter structure
```
 if "greenplum" not in db_version.lower():
   yield
   return
   
gp_recursive_cte_initial = await connection.fetchval("SHOW gp_recursive_cte")
if gp_recursive_cte_initial != "off":
  yield
  return
...
```

File: lib/dl_connector_greenplum/dl_connector_greenplum/core/adapters.py (hunk 2 above)
Comment (KonstantAnxiety) [reply lib/dl_connector_greenplum/dl_connector_greenplum/core/adapters.py:87]: Unfortunately we do
We use postgres in greenplum tests at the moment and I believe that some users rely on this behaviour (not that it is okay, it's not, just saying)

**Review by ovsds — APPROVED** [review]

**Comment by KonstantAnxiety** [comment]
Test fails are not related to the changes in this PR

**Pull request merged**